Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -192,266 +89,700 @@ void LoadSettings()
+		idToDockedInfoMap.erase(dockedClientID);
+	}
+	else
+	{
+		//player offline, edit their character file to put them on last docked solar
+		CAccount* acc = HkGetAccountByCharname(dockedShipName);
+		if (acc)
+		{
+			MoveOfflineShipToLastDockedSolar(dockedShipName);
+		}
+	}
 
-	wstring clientName = (const wchar_t*)Players.GetActiveCharacterName(client);
+	wstring& carrierName = nameToDockedInfoMap[dockedShipName].carrierName;
+
+	//remove the docked ship entry from carrier's data
+	if (nameToCarrierInfoMap.count(carrierName))
+	{
+		auto& dockedList = nameToCarrierInfoMap[carrierName].dockedShipList;
+		for (auto& iter = dockedList.begin(); iter != dockedList.end(); iter++)
+		{
+			if (*iter == dockedShipName)
+			{
+				std::lock_guard<std::mutex> saveLock(saveMutex);
+				dockedList.erase(iter);
+				break;
+			}
+		}
+	}
 
-	if (mobiledockClients[client].mobileDocked)
+	uint carrierId = HkGetClientIdFromCharname(carrierName);
+	if (carrierId != -1)
 	{
+		mobiledockClients[carrierId].iDockingModulesAvailable++;
+	}
+	//finally, clear the docked ship info
+	nameToDockedInfoMap.erase(dockedShipName);
 
-		returncode = SKIPPLUGINS;
+	return Success;
+}
 
-		// Set last base to last real base this ship was on. POB support will be added in 0.9.X version.
-		Players[client].iLastBaseID = mobiledockClients[client].iLastBaseID;
+void DockShipOnCarrier(uint dockingID, uint carrierID)
+{
 
-		// Check if carrier died.
-		if (mobiledockClients[client].carrierDied)
+	// The client is free to dock, erase from the pending list and handle
+	mapPendingDockingRequests.erase(dockingID);
+
+	auto dockingName = reinterpret_cast<const wchar_t*>(Players.GetActiveCharacterName(dockingID));
+	bool shipAlreadyDocked = false;
+	if (mobiledockClients[carrierID].iDockingModulesAvailable == 0)
+	{
+		for (wstring dockedName : idToCarrierInfoMap[carrierID]->dockedShipList)
 		{
-			// If died carrier and docked ship are in same system.
-			if (Players[client].iSystemID == mobiledockClients[client].carrierSystem)
+			if (dockedName == dockingName)
 			{
-				// Now that the data is prepared, send the player to the carrier's location
-				undockingShip = mobiledockClients[client];
-
-				// Clear the client out of the mobiledockClients now that it's no longer docked
-				mobiledockClients.erase(client);
+				shipAlreadyDocked = true;
+				break;
 			}
-			// If not - redirect to proxy base in carrier's system.
-			else
-			{
-				wstring scProxyBase = HkGetSystemNickByID(mobiledockClients[client].carrierSystem) + L"_proxy_base";
-				if (pub::GetBaseID(mobiledockClients[client].proxyBaseID, (wstos(scProxyBase)).c_str()) == -4)
-				{
-					PrintUserCmdText(client, L"No proxy base in system detected. Contact a developer about this please. Required base: %ws", scProxyBase);
-					return;
-				}
+		}
 
-				mobiledockClients[client].carrierSystem = -1;
-			}
+		if (!shipAlreadyDocked)
+		{
+			PrintUserCmdText(dockingID, L"Carrier has no free docking capacity");

```

Comment:
Should we send an access denied message or something at this point?

Reply:
Added

Reply:
See above